wodles/aws/aws-s3.py

```diff
@@ -75,145 +76,59 @@
-                 max_queue_buffer, prefix, delete_file):
-        """
-        AWS Bucket constructor.
+    sql_find_table_names = """
+                            SELECT
+                                tbl_name
+                            FROM
+                                sqlite_master
+                            WHERE
+                                type='table';"""
 
-        :param reparse: Wether to parse already parsed logs or not
-        :param access_key: AWS access key id
-        :param secret_key: AWS secret access key
-        :param profile: AWS profile
-        :param iam_role_arn: IAM Role
-        :param bucket: Bucket name to extract logs from
-        :param only_logs_after: Date after which obtain logs.
-        :param skip_on_error: Wether to continue processing logs or stop when an error takes place
-        :param account_alias: Alias of the AWS account where the bucket is.
-        :param max_queue_buffer: Maximum event length
-        :param prefix: Prefix to filter files in bucket
-        :param delete_file: Wether to delete an already processed file from a bucket or not
-        """
+    sql_db_optimize = "PRAGMA optimize;"
+
+    def __init__(self, access_key, secret_key, aws_profile, iam_role_arn,
+             service_name=None, region=None, bucket=None):
         self.wazuh_path = open('/etc/ossec-init.conf').readline().split('"')[1]
         self.wazuh_queue = '{0}/queue/ossec/queue'.format(self.wazuh_path)
         self.wazuh_wodle = '{0}/wodles/aws'.format(self.wazuh_path)
         self.msg_header = "1:Wazuh-AWS:"
-        self.legacy_db_table_name = 'log_progress'
-        self.db_table_name = 'trail_progress'
-        self.db_path = "{0}/s3_cloudtrail.db".format(self.wazuh_wodle)
+        self.client = self.get_client(access_key=access_key, secret_key=secret_key,
+            profile=aws_profile, iam_role_arn=iam_role_arn, service_name=service_name,
+            region_name=region, bucket=bucket)
+        # db_name is an instance variable of subclass
+        self.db_path = "{0}/{1}.db".format(self.wazuh_wodle, self.db_name)
         self.db_connector = sqlite3.connect(self.db_path)
-        self.retain_db_records = 5000
-        self.reparse = reparse
-        self.bucket = bucket
-        self.client = self.get_s3_client(access_key, secret_key, profile, iam_role_arn)
-        self.only_logs_after = datetime.strptime(only_logs_after, "%Y%m%d")
-        self.skip_on_error = skip_on_error
-        self.account_alias = account_alias
-        self.max_queue_buffer = max_queue_buffer
-        self.prefix = prefix
-        self.delete_file = delete_file
+        self.db_cursor = self.db_connector.cursor()
+        if bucket:
+            self.bucket = bucket
 
-    def get_s3_client(self, access_key, secret_key, profile, iam_role_arn):
+    def get_client(self, access_key, secret_key, profile, iam_role_arn, service_name, region_name, bucket):
```

Comment:
When working with buckets multiple regions can be configured, is the `region_name` parameter a list? Does it support multiple regions? Also, you're using a `bucket` argument here but this is an abstract class used for both buckets and services. I'd consider reimplementing this, leaving the common behaviour for both buckets and services here and adding custom behaviour for each one overriding the method.

Reply:
If service objects need a custom behaviour, define that custom behaviour in the service class. An abstract class don't need to know details about its child classes, its the latter who need to define their specific needs.